### PR TITLE
chore: Use an annotation to see if running stages have already been reported

### PR DESCRIPTION
**Pull request by abayer**
#### Submitter checklist

- [x] Change is code complete and matches issue description.
- [ ] Change is covered by existing or new tests.

#### Description

Should improve performance in build controller + Lighthouse a bit.

I'm running this on prod and it looks to make a real difference in performance thanks to getting rid of getting creds from vault and making a GitHub API call for every completed step.

#### Which issue this PR fixes

n/a

### CODE REVIEW

File: pkg/cmd/controller/controller_build.go [hunk 1]
```diff
@@ -1361,10 +1362,12 @@ func toScmStatusAndDescription(activity *v1.PipelineActivity) (string, string) {
 	}
 	stagesByStatus := activity.StagesByStatus()
 
+	runningStages := ""
 	if len(stagesByStatus[v1.ActivityStatusTypeRunning]) > 0 {
+		runningStages = strings.Join(stagesByStatus[v1.ActivityStatusTypeRunning], ",")
 		description = fmt.Sprintf("Pipeline running stage(s): %s", strings.Join(stagesByStatus[v1.ActivityStatusTypeRunning], ", "))
 	}
-	return scmStatus, description
+	return scmStatus, description, runningStages
```
Comment (hferentschik) [inline pkg/cmd/controller/controller_build.go]: how long does this _runningStages_ get? Is there a risk that this exceed the 64 char limit?

Comment (hferentschik) [reply pkg/cmd/controller/controller_build.go]: I might the stages as array/slice!?

File: pkg/cmd/controller/controller_build.go [hunk 2]
```diff
@@ -1342,7 +1343,7 @@ func CreateReportTargetURL(templateText string, params ReportParams) string {
 	return buf.String()
 }
 
-func toScmStatusAndDescription(activity *v1.PipelineActivity) (string, string) {
+func toScmStatusDescriptionRunningStages(activity *v1.PipelineActivity) (string, string, string) {
```
Comment (hferentschik) [inline pkg/cmd/controller/controller_build.go]: Getting close where I would create a struct and return it, three strings :-(

File: pkg/cmd/controller/controller_build.go (hunk 1 above)
Comment (abayer) [reply pkg/cmd/controller/controller_build.go]: So k8s annotations can be absurdly long - 256kb, if I remember correctly - so this isn't a concern. =)

File: pkg/cmd/controller/controller_build.go (hunk 2 above)
Comment (abayer) [reply pkg/cmd/controller/controller_build.go]: That is fair. Lemme do that.

File: pkg/cmd/controller/controller_build.go [hunk 3]
```diff
@@ -1342,29 +1343,42 @@ func CreateReportTargetURL(templateText string, params ReportParams) string {
 	return buf.String()
 }
 
-func toScmStatusAndDescription(activity *v1.PipelineActivity) (string, string) {
-	scmStatus := ""
-	description := ""
+type reportStatusInfo struct {
```
Comment (hferentschik) [inline pkg/cmd/controller/controller_build.go]: 👍

File: pkg/cmd/controller/controller_build.go [hunk 4]
```diff
@@ -1342,29 +1343,42 @@ func CreateReportTargetURL(templateText string, params ReportParams) string {
...
+		info.description = "Pipeline failed"
 	}
 	stagesByStatus := activity.StagesByStatus()
 
 	if len(stagesByStatus[v1.ActivityStatusTypeRunning]) > 0 {
-		description = fmt.Sprintf("Pipeline running stage(s): %s", strings.Join(stagesByStatus[v1.ActivityStatusTypeRunning], ", "))
+		info.runningStages = strings.Join(stagesByStatus[v1.ActivityStatusTypeRunning], ",")
+		info.description = fmt.Sprintf("Pipeline running stage(s): %s", strings.Join(stagesByStatus[v1.ActivityStatusTypeRunning], ", "))
+		if len(info.description) > 63 {
+			info.description = info.description[:59] + "..."
```
Comment (hferentschik) [inline pkg/cmd/controller/controller_build.go]: :-), I guess not needed after all

File: pkg/kube/constants.go [hunk 5]
```diff
@@ -284,6 +284,8 @@ const (
 	AnnotationGitURLs = "jenkins.io/git-urls"
 	// AnnotationGitReportState used to annotate what state has been reported to git
 	AnnotationGitReportState = "jenkins.io/git-report-state"
+	// AnnotationGitReportRunningStages used to annotate what stages were last reported to git as running
+	AnnotationGitReportRunningStages = "jenkins.io/git-report-running-stages"
```
Comment (hferentschik) [inline pkg/kube/constants.go]: 👍

**Pull request merged**